src/libstore/misc.cc

```diff
@@ -200,6 +200,40 @@ void Store::queryMissing(const std::vector<DerivedPath> & targets,
             auto drv = make_ref<Derivation>(derivationFromPath(bfd.drvPath));
             ParsedDerivation parsedDrv(StorePath(bfd.drvPath), *drv);
 
+            if (!knownOutputPaths && settings.useSubstitutes && parsedDrv.substitutesAllowed()) {
+                // Try to realise the static output hashes.
+
+                // TODO: I'm not sure this logic is quite right, how do we
+                // ensure all the needed paths are present?
+
+                // TODO: Should we also do this->queryRealisation?
+
+                // TODO: Originally, this was only done if the drv was a CA drv.
+                // But you could have a non-CA drv with a CA input. So it may be
+                // necessary to gate this on the CA experimental feature?
+                auto outputHashes = staticOutputHashes(*this, *drv);
+                knownOutputPaths = true;
+
+                for (auto [outputName, hash] : outputHashes) {
+                    if (!bfd.outputs.contains(outputName))
+                        continue;
+
+                    bool found = false;
+                    for (auto &sub : getDefaultSubstituters()) {
+                        auto realisation = sub->queryRealisation({hash, outputName});
+                        if (!realisation)
+                            continue;
+                        invalid.insert(realisation->outPath);
+                        found = true;
+                    }
+                    if (!found) {
+                        knownOutputPaths = false;
+                        break;
+                    }
```

Comment:
Could this run in parallel (using the thread pool we already have)?

Reply:
I thought about this for a moment. There is already a level of paralleism at play, which is `doPath` is running in a thread along other `doPath`s. The paralellism benefit would be if each output's static output has could be queried in parallel for realisation. My following reasoning assumes I've understood the purpose of what you're suggesting.

I can see the potential benefit of that. I have derivations with hundreds of outputs. I like the software going fast.

However, if any of those outputs fail to realise, we must build the derivation. We potentially put in flight a lot of useless work in the negative case. In practice for most derivations, there are only a handful of outputs. Processing the static output hashes in serial allows us to stop early, and stopping early would save requesting the realisations when we have to build anyway. The most time that will be saved in the typical positive case would be 'O(typical # outputs)' of round trips, i.e. a few.

I also observe that for my usecase of tens of derivation outputs, the performance is not unreasonable as it stands.

Any other thoughts here?